File: Searching/Methods/Binary_Search.cpp
#include <bits/stdc++.h>
using namespace std;
int binarySearch(int arr[], int l, int r, int key)
{
    if (l <= r)
    {
        int mid = l + (r - l) / 2;
        if (arr[mid] == key)
        {
            return mid;
        }
        else if (arr[mid] > key)
        {
            r = mid - 1;
            binarySearch(arr, l, r, key);
        }
        else
        {
            l = mid + 1;
            binarySearch(arr, l, r, key);
        }
    }
    return -1;
}

int main()
{
    int arr[] = {1, 2, 3, 4, 5, 6, 7, 8, 9, 10};
    int n = sizeof(arr) / sizeof(arr[0]);
    int key = 11;
    int index = binarySearch(arr, 0, n - 1, key);
    if (index != -1)
    {
        cout << "Found at index " << index << endl;
    }
    else
    {
        cout << "Not found" << endl;
    }
    return 0;
}

File: Java DSA Bootcamp/src/Recursion/Introduction/BinarySearch.java
package Recursion.Introduction;

import java.util.Scanner;

public class BinarySearch {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        int arr[] = new int[n];
        for (int i=0; i<n; i++) {
            arr[i] = sc.nextInt();
        }
        int target = sc.nextInt();
        int ans = binarySearch(arr, target, 0, arr.length-1);
        if (ans == -1) {
            System.out.println("Element not found.");
        } else {
            System.out.println("Element found at index: " + ans);
        }
    }

    public static int binarySearch(int[] arr, int target, int start, int end) {
        if (start >= end) {
            return -1;
        }
        int mid = start + (end - start)/2;
        if (arr[mid] == target) {
            return mid;
        }

        int index = -1;

        if (arr[mid] > target) {
            index = binarySearch(arr, target, start, mid - 1);
        } else {
            index = binarySearch(arr, target, mid + 1, end);
        }
        return index;
    }
}
